x/lockup/keeper/lock.go

```diff
@@ -722,3 +722,78 @@ func (k Keeper) getCoinsFromLocks(locks []types.PeriodLock) sdk.Coins {
 	}
 	return coins
 }
+
+func (k Keeper) RebondTokens(ctx sdk.Context, lockID uint64, owner sdk.AccAddress, coins sdk.Coins) error {
+	lock, err := k.GetLockByID(ctx, lockID)
+	if err != nil {
+		return err
+	}
+
+	if lock.Owner != owner.String() {
+		return fmt.Errorf("lock %d is not owned by %s", lockID, owner)
+	}
+
+	if !lock.IsUnlocking() {
+		return fmt.Errorf("lock %d is not unlocking, rebonding only possible in unlocking stage", lockID)
+	}
+
+	// If all checks pass, we can rebond the tokens
+	err = k.rebondTokens(ctx, owner, *lock, coins)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// rebondTokens rebonds the specified amount of coins from the lock. If nil, rebonds all tokens.
+func (k Keeper) rebondTokens(ctx sdk.Context, owner sdk.AccAddress, lock types.PeriodLock, coins sdk.Coins) error {
+	var rebondedLock types.PeriodLock
+	var err error
+	if coins != nil && !coins.IsEqual(lock.Coins) {
+		// This branch implies we will end up with two locks:
+		// - The first lock will be the original lock with the coins removed, in unlocking state
+		// - The second lock will be the new rebonded lock with the coins added, in not unlocking state
+
+		// remove rebonded lock's coins from original lock
+		lock.Coins = lock.Coins.Sub(coins)
+		err = k.setLock(ctx, lock)
+		if err != nil {
+			return fmt.Errorf("failed to set original lock: %w", err)
+		}
+
+		// create a rebonded lock
+		splitLockID := k.GetLastLockID(ctx) + 1
+		k.SetLastLockID(ctx, splitLockID)
+		rebondedLock = types.NewPeriodLock(splitLockID, owner, lock.Duration, time.Time{}, coins)
+
+		err = k.setLock(ctx, rebondedLock)
+		if err != nil {
+			return fmt.Errorf("failed to set rebonded lock: %w", err)
+		}
+	} else {
+		// This branch implies that we need to completely replace the old lock in state with a new rebonded lock
+
+		// Restart lock timer and set back to the store
+		// Rebonded lock is the same lock as the original lock, but with an empty EndTime.
+		rebondedLock = types.NewPeriodLock(lock.ID, owner, lock.Duration, time.Time{}, lock.Coins)
+		err = k.setLock(ctx, rebondedLock)
+		if err != nil {
+			return err
+		}
+
+		// Remove original lock's refs from state
+		err = k.deleteLockRefs(ctx, types.KeyPrefixUnlocking, lock)
```

Comment:
Question: Do we not need to use the following?
```suggestion
	err = k.deleteLockRefs(ctx, unlockingPrefix(lock.IsUnlocking()), lock)
```

Reply:
not really, because we are already assured at this point that the lock is unlocking, but we can still add that, but currently there will be no purpose to that